### PR TITLE
[19791] Type system only supports single inheritance

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
XTypes v1.3 Clause 7.2.2.4.5

> The Type System supports single inheritance of Aggegated [sic] Types.

IDL v1.4 Clause 7.4.13.4

> Additions to structure definition in order to support single inheritance and void content (no members).

### CODE REVIEW

**Review by richiware — CHANGES_REQUESTED** [review]
Also we have to do the same with bitsets

File: src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java [hunk 1]
```diff
@@ -89,24 +90,23 @@ public boolean isIsType_a()
...
+        }
+        else if (superType_ != null)
+        {
+            throw new RuntimeGenerationException("Only single type inheritance is supported");
         }
     }
 
     @Override
-    public ArrayList<TypeCode> getInheritances()
+    public TypeCode getInheritance()
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java:107]: The API has already been broken in this previous PR: #94 :smirk:

File: src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java [hunk 2]
```diff
@@ -89,24 +90,23 @@ public boolean isIsType_a()
...
-        if (parent instanceof StructTypeCode)
+        if (superType_ == null && parent instanceof StructTypeCode)
         {
             StructTypeCode parent_struct = (StructTypeCode)parent;
-            superTypes_.add(parent_struct);
+            superType_ = parent_struct;
+        }
+        else if (superType_ != null)
+        {
+            throw new RuntimeGenerationException("Only single type inheritance is supported");
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java:102]: IDL v1.4 Clause 7.4.13.4.1

> Single inheritance is denoted by a colon (:) followed by a scoped name that must correspond to the name of a
previously defined structure.

I suppose the only other possible case would be if it is an alias. Should we consider that case?

[the remark this answers is not in the document]
Comment (richiware) [reply src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java:102]: Yes, and I suggest including that case in `inheritance.idl` file in `dds-types-test` repo.

[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java:102]: Use ParseException: 7a3ad29
Structure alias and other types: 31ed795
Test case:

* eProsima/dds-types-test#11

File: src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java [hunk 3]
```diff
@@ -215,7 +203,7 @@ public void addAnnotation(
         super.addAnnotation(ctx, annotation);
     }
 
-    private ArrayList<StructTypeCode> superTypes_;
+    private StructTypeCode superType_;
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/main/java/com/eprosima/idl/parser/typecode/StructTypeCode.java:215]: Done in 5a5ad74

**Comment by JLBuenoLopez-eProsima** [comment]
Bitsets done in 66f7590

File: src/main/java/com/eprosima/idl/parser/typecode/BitsetTypeCode.java [hunk 4]
```diff
@@ -116,15 +114,36 @@ public boolean addBitfield(
...
+            BitsetTypeCode parent_bitset = (BitsetTypeCode)parent.getContentTypeCode();
+            super_type_ = parent_bitset;
+        }
+        else if (super_type_ != null)
+        {
+            throw new ParseException("Only single type inheritance is supported");
+        }
+        else
+        {
+            throw new ParseException("Inheritance must correspond to the name of a previously defined bit set");
```
Comment (richiware) [inline src/main/java/com/eprosima/idl/parser/typecode/BitsetTypeCode.java:139]:
```suggestion
            throw new ParseException("Inheritance must correspond to the name of a previously defined bitset");
```

**Review by richiware — CHANGES_REQUESTED** [review]

File: src/main/java/com/eprosima/idl/parser/typecode/BitsetTypeCode.java (hunk 4 above)
Comment (JLBuenoLopez-eProsima) [reply src/main/java/com/eprosima/idl/parser/typecode/BitsetTypeCode.java]: Done in e903808

**Review by richiware — APPROVED** [review]

**Pull request merged**